### ISSUE TITLE
helidon-microprofile-openapi dependency convergence and duplicate class problems

### ISSUE THREAD

**Issue by Holubow** [oracle/helidon#1559]
## Problem Description
My project uses a handful of Helidon MicroProfile 2.0.0-M2 dependencies and I run the maven-enforcer-plugin and the duplicate-finder-plugin and I have found that helidon-microprofile-openapi has dependency convergence and duplicate class problems with other Helidon modules.

It seems that helidon-microprofile-openapi is using newer versions of some dependencies shared with other Helidon modules.

This is an example of the exclusions I need to apply to helidon-microprofile-openapi in my Maven POM to address the problems for illustration:

	<exclusions>
		
		<exclusion>
			<groupId>org.jboss.logging</groupId>
			<artifactId>jboss-logging</artifactId>
		</exclusion>
		
		<exclusion>
			<groupId>org.jboss</groupId>
			<artifactId>jandex</artifactId>
		</exclusion>
		
		<exclusion>
			<groupId>org.eclipse.microprofile.config</groupId>
			<artifactId>microprofile-config-api</artifactId>
		</exclusion>
		
		<exclusion>
			<groupId>jakarta.json</groupId>
			<artifactId>jakarta.json-api</artifactId>
		</exclusion>
		
		<exclusion>
			<groupId>org.glassfish</groupId>
			<artifactId>jakarta.json</artifactId>
		</exclusion>
	</exclusions>

## Environment Details
* Helidon Version: 2.0.0-M2
* Helidon SE or Helidon MP: Helidon MP
* JDK version: jdk-11.0.5
* OS: win64
* Docker version (if applicable):

----------

## Problem Description
[//]: # "Describe the bug in detail highlighting current behavior vs expected behavior"
[//]: # "State if the problem is easily reproducible or happens intermittently"
[//]: # "Include stack traces or command outputs"

## Steps to reproduce
[//]: # "Step by step instructions to reproduce the problem"
[//]: # "Provide sample code/application if relevant"

Create a Helidon project with dependencies on helidon.microprofile-server and helidon-microprofile-openapi.
Look at the different versions of the dependencies listed above in helidon.microprofile-server and helidon-microprofile-openapi using Eclipse's Dependency Hierarchy view.
The maven-enforcer-plugin will complain about dependency convergence errors.
The duplicate-finder-maven-plugin will complain about duplicate classes.
Reviewing the underlying POM versions should also illustrate the difference in versions.

**Comment by romain-grecourt** [comment]
- Can you share the dependency tree? (`mvn dependency:tree`)
- Do you use the `helidon-mp` parent pom ?

FYI, all Helidon modules share the same dependency management.

**Comment by Holubow** [comment]
Here is the dependency tree for helidon-microprofile-openapi:

```
 --- maven-dependency-plugin:3.0.0:tree (default-cli) @ helidon-microprofile-openapi ---
 io.helidon.microprofile.openapi:helidon-microprofile-openapi:jar:2.0.0-M2
 +- org.eclipse.microprofile.config:microprofile-config-api:jar:1.3:compile
 +- io.helidon.microprofile.server:helidon-microprofile-server:jar:2.0.0-M2:compile
 |  +- io.helidon.microprofile.cdi:helidon-microprofile-cdi:jar:2.0.0-M2:compile
 |  |  +- io.helidon.microprofile.weld:weld-se-core:jar:2.0.0-M2:compile
 |  |  |  +- io.helidon.microprofile.weld:weld-core-impl:jar:2.0.0-M2:compile
 |  |  |  +- org.jboss.weld.environment:weld-environment-common:jar:3.1.2.Final:compile
 |  |  |  +- org.jboss.weld:weld-api:jar:3.1.SP1:compile
 |  |  |  +- org.jboss.weld.probe:weld-probe-core:jar:3.1.2.Final:compile
 |  |  |  +- org.jboss.weld:weld-spi:jar:3.1.SP1:compile
 |  |  |  +- org.jboss.spec.javax.annotation:jboss-annotations-api_1.3_spec:jar:1.0.0.Final:compile
 |  |  |  +- org.jboss.spec.javax.el:jboss-el-api_3.0_spec:jar:1.0.0.Final:compile
 |  |  |  +- org.jboss.spec.javax.interceptor:jboss-interceptors-api_1.2_spec:jar:1.0.0.Final:compile
 |  |  |  +- javax.enterprise:cdi-api:jar:2.0:compile
 |  |  |  +- javax.inject:javax.inject:jar:1:compile
 |  |  |  \- org.jboss.classfilewriter:jboss-classfilewriter:jar:1.2.4.Final:compile
 |  |  \- io.helidon.common:helidon-common-context:jar:2.0.0-M2:compile
 |  +- io.helidon.webserver:helidon-webserver:jar:2.0.0-M2:compile
 |  |  +- io.helidon.common:helidon-common-reactive:jar:2.0.0-M2:compile
 |  |  |  \- io.helidon.common:helidon-common-mapper:jar:2.0.0-M2:compile
 |  |  +- io.helidon.common:helidon-common:jar:2.0.0-M2:compile
 |  |  +- io.helidon.common:helidon-common-http:jar:2.0.0-M2:compile
 |  |  +- io.helidon.media:helidon-media-common:jar:2.0.0-M2:compile
 |  |  +- io.helidon.common:helidon-common-key-util:jar:2.0.0-M2:compile
 |  |  |  \- io.helidon.common:helidon-common-configurable:jar:2.0.0-M2:compile
 |  |  +- io.opentracing:opentracing-util:jar:0.33.0:compile
 |  |  +- io.opentracing:opentracing-api:jar:0.33.0:compile
 |  |  +- io.opentracing:opentracing-noop:jar:0.33.0:compile
 |  |  +- io.helidon.tracing:helidon-tracing-config:jar:2.0.0-M2:compile
 |  |  +- io.netty:netty-codec-http:jar:4.1.45.Final:compile
 |  |  |  +- io.netty:netty-common:jar:4.1.45.Final:compile
 |  |  |  +- io.netty:netty-buffer:jar:4.1.45.Final:compile
 |  |  |  +- io.netty:netty-transport:jar:4.1.45.Final:compile
 |  |  |  |  \- io.netty:netty-resolver:jar:4.1.45.Final:compile
 |  |  |  \- io.netty:netty-codec:jar:4.1.45.Final:compile
 |  |  +- io.netty:netty-codec-http2:jar:4.1.45.Final:compile
 |  |  \- io.netty:netty-handler:jar:4.1.45.Final:compile
 |  +- io.helidon.webserver:helidon-webserver-jersey:jar:2.0.0-M2:compile
 |  |  \- io.helidon.jersey:helidon-jersey-server:jar:2.0.0-M2:compile
 |  |     +- org.glassfish.jersey.core:jersey-server:jar:2.30.1:compile
 |  |     |  +- org.glassfish.jersey.core:jersey-common:jar:2.30.1:compile
 |  |     |  |  \- org.glassfish.hk2:osgi-resource-locator:jar:1.0.3:compile
 |  |     |  +- org.glassfish.jersey.core:jersey-client:jar:2.30.1:compile
 |  |     |  \- jakarta.validation:jakarta.validation-api:jar:2.0.2:compile
 |  |     +- org.glassfish.jersey.inject:jersey-hk2:jar:2.30.1:compile
 |  |     |  +- org.glassfish.hk2:hk2-locator:jar:2.6.1:compile
 |  |     |  |  +- org.glassfish.hk2.external:aopalliance-repackaged:jar:2.6.1:compile
 |  |     |  |  +- org.glassfish.hk2:hk2-api:jar:2.6.1:compile
 |  |     |  |  \- org.glassfish.hk2:hk2-utils:jar:2.6.1:compile
 |  |     |  \- org.javassist:javassist:jar:3.25.0-GA:compile
 |  |     +- javax.xml.bind:jaxb-api:jar:2.3.0:compile
 |  |     \- javax.activation:javax.activation-api:jar:1.2.0:compile
 |  +- io.helidon.common:helidon-common-service-loader:jar:2.0.0-M2:compile
 |  |  \- javax.annotation:javax.annotation-api:jar:1.3.1:compile
 |  +- org.glassfish:javax.json:jar:1.1.2:compile
 |  |  \- javax.json:javax.json-api:jar:1.1.2:compile
 |  +- io.helidon.microprofile.config:helidon-microprofile-config:jar:2.0.0-M2:runtime
 |  |  \- io.helidon.config:helidon-config-object-mapping:jar:2.0.0-M2:runtime
 |  +- io.helidon.media.jsonp:helidon-media-jsonp-common:jar:2.0.0-M2:compile
 |  +- io.helidon.config:helidon-config-yaml:jar:2.0.0-M2:runtime
 |  +- org.slf4j:slf4j-jdk14:jar:1.7.26:runtime
 |  |  \- org.slf4j:slf4j-api:jar:1.7.26:runtime
 |  +- org.glassfish.jersey.ext.cdi:jersey-weld2-se:jar:2.30.1:runtime
 |  |  +- org.glassfish.jersey.ext.cdi:jersey-cdi1x:jar:2.30.1:runtime
 |  |  \- jakarta.ws.rs:jakarta.ws.rs-api:jar:2.1.6:compile
 |  \- io.helidon.jersey:helidon-jersey-media-jsonp:jar:2.0.0-M2:runtime
 |     \- org.glassfish.jersey.media:jersey-media-json-processing:jar:2.30.1:runtime
 |        \- org.glassfish:jsonp-jaxrs:jar:1.1.5:runtime
 +- io.helidon.openapi:helidon-openapi:jar:2.0.0-M2:compile
 |  +- io.smallrye:smallrye-open-api:jar:1.2.0:compile
 |  |  \- org.jboss.logging:jboss-logging:jar:3.2.1.Final:compile
 |  +- io.helidon.media.jsonp:helidon-media-jsonp-server:jar:2.0.0-M2:compile
 |  +- org.eclipse:yasson:jar:1.0.6:compile
 |  |  +- jakarta.json.bind:jakarta.json.bind-api:jar:1.0.2:compile
 |  |  +- jakarta.json:jakarta.json-api:jar:1.1.6:compile
 |  |  \- org.glassfish:jakarta.json:jar:module:1.1.6:compile
 |  +- io.helidon.config:helidon-config:jar:2.0.0-M2:compile
 |  |  \- io.helidon.common:helidon-common-media-type:jar:2.0.0-M2:compile
 |  +- org.jboss:jandex:jar:2.1.1.Final:compile
 |  \- org.eclipse.microprofile.openapi:microprofile-openapi-api:jar:1.1.2:compile
 +- javax.interceptor:javax.interceptor-api:jar:1.2.2:provided
 +- org.junit.jupiter:junit-jupiter-api:jar:5.1.0:test
 |  +- org.apiguardian:apiguardian-api:jar:1.0.0:test
 |  +- org.opentest4j:opentest4j:jar:1.0.0:test
 |  \- org.junit.platform:junit-platform-commons:jar:1.1.0:test
 +- org.junit.jupiter:junit-jupiter-params:jar:5.1.0:test
 +- org.junit.jupiter:junit-jupiter-engine:jar:5.1.0:test
 |  \- org.junit.platform:junit-platform-engine:jar:1.1.0:test
 \- org.yaml:snakeyaml:jar:1.24:test
```

Search for "json-api" in the helidon-microprofile-openapi dependency tree - you should see two versions that include duplicate classes in the JAR: javax.json:javax.json-api:jar:1.1.2:compile and jakarta.json:jakarta.json-api:jar:1.1.6:compile (from org.eclipse.yasson).

Search for ".json:jar" in the helidon-microprofile-openapi dependency tree - you should see two versions that include duplicate classes in the JAR: org.glassfish:javax.json:jar:1.1.2:compile and org.glassfish:jakarta.json:jar:module:1.1.6:compile (from org.eclipse.yasson).

Here are dependency trees included in the warnings from maven-enforcer-plugin coming from io.smallrye:smallrye-open-api:

```
  +-io.helidon.microprofile.openapi:helidon-microprofile-openapi:2.0.0-M2
    +-io.helidon.openapi:helidon-openapi:2.0.0-M2
      +-io.smallrye:smallrye-open-api:1.2.0
        +-org.eclipse.microprofile.config:microprofile-config-api:1.4
```

Helidon projects use org.eclipse.microprofile.config:microprofile-config-api:1.3 instead.

```
  +-io.helidon.microprofile.openapi:helidon-microprofile-openapi:2.0.0-M2
    +-io.helidon.openapi:helidon-openapi:2.0.0-M2
      +-io.smallrye:smallrye-open-api:1.2.0
        +-org.jboss.logging:jboss-logging:3.4.1.Final
```

Helidon projects use org.jboss.logging:jboss-logging:3.2.1.Final instead.

```
  +-io.helidon.microprofile.openapi:helidon-microprofile-openapi:2.0.0-M2
    +-io.helidon.openapi:helidon-openapi:2.0.0-M2
      +-io.smallrye:smallrye-open-api:1.2.0
        +-org.jboss:jandex:2.1.2.Final
```


Helidon projects use org.jboss:jandex:2.1.1.Final instead.

**Comment by romain-grecourt** [comment]
Is this the dependency tree of the `helidon-microprofile-openapi` module ?
Please include a dependency tree of your project. Does it inherit the `helidon-mp` parent pom ?

The class-path duplicates between `javax` and `jakarta` is a never ending battle. CC @tomas-langer. This will end when we switch over to `jakarta`. We may have some exclusions somewhere.

**Comment by tjquinno** [comment]
Update: I have been using the MP quickstart app as a testbed, and running the enforcer on that project. Doing so I was able to reproduce the errors reported in this issue.

With some advice from Romain, I resolved the dependency convergence issue for JBoss logging  by moving its declarations from the top-level `pom.xml` to `helidon-dependencies/pom.xml`.

Moving the declarations for `org.glassfish:jakarta.json`, though, did not resolve that one. Still working.

**Comment by Holubow** [comment]
I did provide the dependency tree of the helidon-microprofile-openapi module that I collected when building it.
My project does not inherit the helidon-mp parent POM and it is irrelevant as I have illustrated the problem in Helidon. My project can workaround these dependency problems via exclusions, but it would be nice if I didn't have to if the dependency issues could be addressed on the Helidon side.
